pkg/controller/acmeorders/sync.go

```diff
@@ -461,90 +419,157 @@ func determineSolverConfigToUse(candidates []cmapi.ACMEChallengeSolver, authz *a
 		return nil
 	}
 
-	// this variable tracks the number of labels that matched when a solver
-	// that specifically names the dnsName in the authorization matches.
-	// This is used to tie-break if two different solver configurations both
-	// explicitly name a dnsName
-	numLabelsSpecificMatch := 0
-	var specificMatch *cmapi.ACMEChallengeSolver
-	var specificMatchToSolve *acmeapi.Challenge
-
-	// this variable tracks the number of labels that matched when a solver
-	// that does NOT specifically list the authorization's dnsName matches.
-	// If no solver explicitly lists the dnsName then the solver that matches
-	// the most labels is used.
-	matchAllDomainsNumLabels := 0
-	// matchAll is the most-specific solver that matches the authorization,
-	// that does not list the authorization's dns name
-	var matchAll *cmapi.ACMEChallengeSolver
-	var matchAllToSolve *acmeapi.Challenge
-
-	for idx := range candidates {
-		d := &candidates[idx]
-		acmech := challengeForSolver(d)
+	// 2. filter solvers to only those that matchLabels
+	for _, cfg := range solvers {
```

Comment:
This loop is quite big, splitting some of it into function calls would provide a clearer way of understanding how this filtering works

Reply:
I've had a go at splitting this into separate function calls, however the nature of this selection logic is quite interdependent between the different switch statements, e.g. if two selectors *both* match dnsNames, we have to compare the label & dnsZone selectors to determine whether this solver should be chosen.

Making this a separate function would require almost all of the same logic from this function to be copied across into the new one. I think we can probably explore ways to refactor selector matching/resolution in future, but right now I think this is a net-improvement over the previous logic (which is even more confusing 😓).